### PR TITLE
Add LCA

### PR DESCRIPTION
**Pull request by didwhddks**
- [x] Read [Contribute to NTHU-CPP](https://nthu-cp.github.io/NTHU-CPP/others/contribution.html).
- [x] Rebase to the latest main branch.
- [x] List all the references.
- [x] Successfully build the website by mdBook with no error after the change.
- [x] Exclude any irrelevant files.
- [x] Link to the issue which is related to your change.
- [x] Review the content by myself.
- [x] Pass tests/CI.

### CODE REVIEW

**Review by harry900831 — CHANGES_REQUESTED** [review]
大致瀏覽了一下，幾點建議
- 新增對高度分塊的LCA 作法，這裡面有介紹。https://www.topcoder.com/thrive/articles/Range%20Minimum%20Query%20and%20Lowest%20Common%20Ancestor
- 需要簡介 LCA 的性質，以及最後提供一些例題能使用 LCA去解。
- 在說明可以將 LCA 轉換成 RMQ 問題後，不需要針對怎麼做 RMQ 問題有過多著墨，未來可能會有 RMQ 問題的文章，可以直接 refer 到那邊

我先將此 PR 轉成 draft PR，等你完成時再轉成 ready for review，到時我會再更詳細的去review，謝謝。

**Comment by didwhddks** [comment]
@harry900831 助教你好，我已經大致完成了文章的初步架構，目前還缺 LCA 的例題以及高度分塊 LCA 的 sample code。想先請助教幫我 review 一下目前的文章內容。

File: src/tree/lowest_common_ancestor.md [hunk 1]
```diff
@@ -0,0 +1,730 @@
...
+在以上兩種 Case 當中，可以發現點 \\( u \\) 所處並查集的祖先節點均為所求點對 \\( (u,\ v) \\) 之最近共同祖先，因此 Tarjan 的想法是正確的。至於為什麼此演算法強制離線，因為 Tarjan 的想法是在 DFS 的走訪過程中去觀察子樹的結構。如果事先不知道查詢點對為何，我們就無法在走訪過程中去檢查點對之間的關係。
+
+總結一下此作法的時間複雜度：
+
+- 初始化每個點的並查集和每個集合的祖先節點：\\( O(N) \\)
+- DFS 走訪所有點：\\( O(N) \\)
+- 每當一點 \\( v \\) 搜索完其所有子樹節點時，檢查所有形式為 \\( (u,\ v) \\) 的查詢點對，點 \\( u \\) 是否為搜索過的狀態。如果已搜索過，則當前查詢點對的最近共同祖先為點 \\( u \\) 所處並查集的祖先節點：\\( O(Q\alpha(N)) \\) --- 每個查詢點對只會被檢查兩次。
+- 接著返回其父節點 \\( p \\) 時，將點 \\( v \\) 所處集合和點 \\( p \\) 所處集合合併，並將合併集合的祖先節點設為點 \\( p \\)：\\( O(N\alpha(N)) \\) --- 至多合併 \\( N \\) 次。
+
+以上所有 DSU 的操作都可以用 union-by-rank 和 path compression 來優化。因此針對離線的查詢，Tarjan's algorithm 可以有效率地在 \\( O((N+Q)\alpha(N)) \\) 的時間內完成，相當趨近於線性時間。
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:262]: DSU -> Disjoint Set

Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:262]: 你這句話應該在時間複雜度分析之前提

File: src/tree/lowest_common_ancestor.md [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+  return 0;
+}
+
+```
+
+</details>
+
+### Heavy Path Decomposition
+
+此小節主要關注於如何利用樹重鏈剖分的概念來查詢點對的最近共同祖先，如果讀者對於樹重鏈的定義以及其相關操作（跳輕邊）還不太清楚的話，建議可以先去觀看相關章節。
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:356]: Here should be a link to https://github.com/NTHU-CP/NTHU-CPP/pull/62.
Maybe we could add this link after HLD be merged.

File: src/tree/lowest_common_ancestor.md [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+
+```
+
+</details>
+
+### 轉化成 RMQ 問題
+
+有許多問題可以透過轉化成其他問題的方式來獲得很好的解法，LCA 問題也不是例外。我們可以透過尤拉路徑在線性時間內將 LCA 問題轉化成另一經典問題（Range Minimum Query），解決 RMQ 問題的同時我們也就解決了 LCA 問題。
+
+#### 建立尤拉路徑
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:479]: Eulerian path
尤拉 或 歐拉
可能可以跟讀者說一下兩種翻譯都有人用

File: src/tree/lowest_common_ancestor.md [hunk 4]
```diff
@@ -0,0 +1,730 @@
...
+
+我們可以利用 DFS 走訪來建立尤拉路徑，通常會用一陣列來儲存。在 DFS 的走訪過程中，每經過一個點我們就把該點加入到尤拉路徑，包含返回時所經過的點，不是很清楚的讀者可以看以下示意圖：
+
+<img src="image/LCA/RMQ_1.png" width="700" style="display:block; margin: 0 auto;"/>
+
+觀察尤拉路徑我們可以發現，假設我們想知道點對 \\( (u,\ v) \\) 的最近共同祖先是誰，我們只要沿著尤拉路徑從點 \\( u \\) 到點 \\( v \\) 找路徑上深度最淺的點就會是點對 \\( (u,\ v) \\) 的 LCA，因為從點 \\( u \\) 到點 \\( v \\) 必定會先返回其 LCA，然後再到點 \\( v \\)，在這之間也不會出現比其 LCA 深度更淺的點。比如說如下圖所示，點對 \\( (1,\ 8) \\) 的最近共同祖先就會是從點 \\( 1 \\) 到點 \\( 8 \\) 路徑上深度最淺的點，也就是點 \\( 6 \\)。
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:485]: u, v 可能會出現很多次->取第一次出現即可

Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:485]: 正確性需要說明

Comment (harry900831) [reply src/tree/lowest_common_ancestor.md:485]: 包含為什麼取第一次就好

File: src/tree/lowest_common_ancestor.md [hunk 5]
```diff
@@ -0,0 +1,730 @@
...
+
+有了上述預處理好的陣列後，剩下的工作就會是處理 RMQ 問題。對於 RMQ 問題，有許多經典的解法像是線段樹、序列分塊或是 Sparse Table，這邊我們不會細講這些經典解法的概念和實作細節，有興趣的讀者可以去觀看相關的章節。而此小節我們主要關注於如何利用高度分塊來達到 \\( O(N) \\) 預處理、每次查詢 \\( O(1) \\) 的做法。
+
+#### 高度分塊區間最小值查詢
+
+觀察 Depth 陣列我們可以發現，陣列中相鄰的值會剛好差 1。原因是因為在建立尤拉路徑時，我們每次只會往上或是往下走一步，所以在 Euler Tour 陣列中相鄰的點，點深度會剛好差 1。基於此特性，我們可以用來優化區間最小值的查詢。
+
+高度分塊的 RMQ 同樣是基於序列分塊的概念，先將 Depth 陣列分成 \\( \frac{N}{K} \\) 個大小為 \\( K \\) 的塊，每一塊維護該塊最小值所對應到的 index，存放在另開的 Block 陣列中，如下圖所示。
+
+<img src="image/LCA/RMQ_4.png" width="700" style="display:block; margin: 0 auto;"/>
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:501]: 範例中K=2

File: src/tree/lowest_common_ancestor.md [hunk 6]
```diff
@@ -0,0 +1,730 @@
...
+
+對於搜索塊之間的最小值，相較於直接遍歷過 \\( l \\) 和 \\( r \\) 中間所有塊，我們可以在 \\(O\left(\frac{N}{K}\log\frac{N}{K}\right) \\) 的時間內先對 Block 陣列建 Sparse Table。基於 RMQ 的特性，之後的查詢可以在 \\( O(1) \\) 時間內辦到 。而對於在塊內搜索最小值，暴力做的話會需要 \\( O(K) \\) 的時間，有沒有更快的作法？
+
+還記得在此小節一開始我們提及的尤拉路徑的特性嗎？藉由該特性，我們其實可以透過像是建立差分數組的方式來將每一塊 block 轉換成大小為 \\( K-1 \\) 的塊，使得新塊內元素只會有 \\( +1 \\) 和 \\( -1 \\) 兩種可能性。舉例來說，假設存在一塊 block A 其內容為 \\( \left[0,1,2,1,0\right] \\)，我們可以將其轉換成內容為 \\( \left[1,1,-1,-1\right] \\) 的新塊。而一般來說我們會將 \\( -1 \\) 轉成 \\( 0 \\)，用 bitmask 的方式來方便紀錄以及識別每一種新塊。
+
+透過此轉換，我們可以發現新塊最多只會有 \\( 2^{K-1} \\) 種。而轉換後有著相同內容的塊，塊內所有區間組合的最小值在塊內所對應的 index offset 都會是相同的，因為新塊內容其實間接說明了原塊內的深度分布。因此，相較於每次查詢時都去暴力搜索塊內的最小值，我們其實可以預處理每一種新塊內所有區間組合的最小值在塊內所對應的 index offset，這部分會花費 \\( O(2^{K-1}K^2) \\) 的時間。針對之後塊內區間搜索最小值的查詢，我們就可以在 \\( O(1) \\) 的時間內查表得知。
+
+為了比較好的時間複雜度，一般我們會取 \\( K \\) 等於 \\( 0.5\log N \\)，這樣我們就可以把預處理的時間複雜度壓到 \\( O(N) \\)。
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:511]: why 0.5 log N?
can't K = logN get the same O(N) - O(1) either?

File: src/tree/lowest_common_ancestor.md [hunk 7]
```diff
@@ -0,0 +1,730 @@
...
+
+總結一下此作法跟整體時間複雜度：
+
+- 預處理
+  - DFS 一次建立尤拉路徑，同時記錄 Depth 和 First Occurrence 陣列：\\( O(N) \\)。
+
+  - 將 Depth 陣列分塊，每一塊維護最小值所對應的 index 並存放在 Block 陣列。與此同時，計算每一塊的 bitmask 並存放在另開的 Mask 陣列裡：\\( O(N) \\)。
+
+  - 對 Block 陣列建 Sparse Table：\\(O\left(\frac{N}{K}\log\frac{N}{K}\right) \xrightarrow{K=0.5\log N} O(N) \\)
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:520]: 稍微推導一下吧

File: src/tree/lowest_common_ancestor.md [hunk 8]
```diff
@@ -0,0 +1,730 @@
...
+
+<img src="image/LCA/RMQ_2.png" width="700" style="display:block; margin: 0 auto;"/>
+
+因此，透過尤拉路徑我們可以將 LCA 問題轉化成找區間最小值的問題。一般來說，我們會另開一陣列來記錄每個點在尤拉路徑上第一次出現時所對應的 index，方便我們在 \\( O(1) \\) 時間內找到對應的區間。
+
+<img src="image/LCA/RMQ_3.png" width="700" style="display:block; margin: 0 auto;"/>
+
+有了上述預處理好的陣列後，剩下的工作就會是處理 RMQ 問題。對於 RMQ 問題，有許多經典的解法像是線段樹、序列分塊或是 Sparse Table，這邊我們不會細講這些經典解法的概念和實作細節，有興趣的讀者可以去觀看相關的章節。而此小節我們主要關注於如何利用高度分塊來達到 \\( O(N) \\) 預處理、每次查詢 \\( O(1) \\) 的做法。
+
+#### 高度分塊區間最小值查詢
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:495]: I think O(n) -- O(1) LCQ is a better title

File: src/tree/lowest_common_ancestor.md [hunk 9]
```diff
@@ -0,0 +1,730 @@
...
+    int u, v;
+    cin >> u >> v;
+    cout << LCA(u, v) << "\n";
+  }
+  return 0;
+}
+
+```
+
+</details>
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:675]: 最後補一句：這個方法常數很高 比賽中不實用 但是個很漂亮的做法達到理論最佳 online O(n) O(1)

File: src/tree/lowest_common_ancestor.md [hunk 10]
```diff
@@ -0,0 +1,730 @@
...
+> - 給定 \\( a \\) 和 \\( b \\)，請輸出樹上有多少點到點 \\( a \\) 的距離和到點 \\( b \\) 的距離相同。
+>
+> - \\( N,\ Q \leq 10^5 \\)
+
+<details><summary> Solution </summary>
+
+觀察一下可以發現，我們可以在點 \\( a \\) 到點 \\( b \\) 的最短路徑上找中點。如果找不到中點（點 \\( a \\) 到點 \\( b \\) 的距離為奇數），代表樹上沒有任何點滿足題目要求。而如果找的到中點，我們可以分成兩種 Case 來討論。Case 1 是中點剛好為 \\( LCA(a,\ b) \\) 的情況，
+
+</details>
+
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:708]: https://codeforces.com/contest/609/problem/E

In my opinion, you can add this problem as an exercise, to teach the reader that you can use binary lifting to maintain the maximum weight edge on a tree path.

Comment (harry900831) [reply src/tree/lowest_common_ancestor.md:708]: (It would be great if you could provide code)

**Review by harry900831 — CHANGES_REQUESTED** [review]
Overall, you did a great job!
Please fix textlint, thanks.

File: src/tree/lowest_common_ancestor.md [hunk 11]
```diff
@@ -0,0 +1,730 @@
...
+對於搜索塊之間的最小值，相較於直接遍歷過 \\( l \\) 和 \\( r \\) 中間所有塊，我們可以在 \\(O\left(\frac{N}{K}\log\frac{N}{K}\right) \\) 的時間內先對 Block 陣列建 Sparse Table。基於 RMQ 的特性，之後的查詢可以在 \\( O(1) \\) 時間內辦到 。而對於在塊內搜索最小值，暴力做的話會需要 \\( O(K) \\) 的時間，有沒有更快的作法？
+
+還記得在此小節一開始我們提及的尤拉路徑的特性嗎？藉由該特性，我們其實可以透過像是建立差分數組的方式來將每一塊 block 轉換成大小為 \\( K-1 \\) 的塊，使得新塊內元素只會有 \\( +1 \\) 和 \\( -1 \\) 兩種可能性。舉例來說，假設存在一塊 block A 其內容為 \\( \left[0,1,2,1,0\right] \\)，我們可以將其轉換成內容為 \\( \left[1,1,-1,-1\right] \\) 的新塊。而一般來說我們會將 \\( -1 \\) 轉成 \\( 0 \\)，用 bitmask 的方式來方便紀錄以及識別每一種新塊。
+
+透過此轉換，我們可以發現新塊最多只會有 \\( 2^{K-1} \\) 種。而轉換後有著相同內容的塊，塊內所有區間組合的最小值在塊內所對應的 index offset 都會是相同的，因為新塊內容其實間接說明了原塊內的深度分布。因此，相較於每次查詢時都去暴力搜索塊內的最小值，我們其實可以預處理每一種新塊內所有區間組合的最小值在塊內所對應的 index offset，這部分會花費 \\( O(2^{K-1}K^2) \\) 的時間。針對之後塊內區間搜索最小值的查詢，我們就可以在 \\( O(1) \\) 的時間內查表得知。
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:509]: 舉個例吧

**Review by harry900831 — CHANGES_REQUESTED** [review]

File: src/tree/lowest_common_ancestor.md [hunk 12]
```diff
@@ -0,0 +1,928 @@
...
+>
+> - \\( N,\ Q \leq 2 \times 10^5 \\)
+
+基本上此問題就是在問樹上兩點的最近共同祖先為何，下面我們將介紹幾種常見的作法。
+
+## 演算法
+
+### Brute Force
+
+一個直觀的做法是先從根開始作一次 DFS，確認出每個點的父節點以及時間戳（第一次訪問與最後一次訪問的時間）。接著針對詢問，我們可以藉由預處理好的的資訊暴力往上搜索所求點對 \\( (u,\ v) \\) 中深度較淺的點的祖先，直到當前搜索的點同時也為另一點的祖先，該點即為點對 \\( (u,\ v) \\) 的最近共同祖先。而確認一個點是否為另一點的祖先，我們可以利用時間戳在 \\( O(1) \\) 的時間內得知。
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:38]: > 第一次訪問與最後一次訪問的時間

DFS 進行遞迴時進入該節點與離開該節點時間戳記

Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:38]: 解釋一下怎麼用時間戳記O(1) 判斷祖先

File: src/tree/lowest_common_ancestor.md [hunk 13]
```diff
@@ -0,0 +1,730 @@
...
+
+根據這個想法，我們可以列出下列遞迴關係式：
+
+$$ancestor(u,\ i) = \begin{cases} parent(u) & \text {if $i = 0$} \newline ancestor(ancestor(u,\ i-1),\ i-1) & \text{if $i > 0$} \end{cases}$$
+
+這邊 \\( ancestor(u,\ i) \\) 代表點 \\( u \\) 的第 \\( 2^i \\) 個祖先。
+
+利用上述關係式解 DP，我們可以在 \\( O(N\log N) \\) 的時間內預處理完所有點的第 \\( 2^i \\) 個祖先的資訊。等到需要進行點的抬升時，我們即可以在 \\( O(\log N) \\) 的時間內完成。
+
+除了抬升的優化以外，對於等深度的兩點找 LCA 本身具有單調性，直覺上可以利用二分搜來加速。但在稍早提到的 Brute force 當中，我們並沒有提到相關的操作。原因是因為我們只記錄了每個點的父節點的資訊，在處理二分搜的檢查函數上會花費線性的時間，導致整個二分搜的時間複雜度為 \\( O(N\log N) \\)，明顯沒有比較快。
```
[the remark this answers is not in the document]
Comment (harry900831) [reply src/tree/lowest_common_ancestor.md]: 還是沒有解釋怎麼二分艘

File: src/tree/lowest_common_ancestor.md (hunk 8 above)
Comment (harry900831) [reply src/tree/lowest_common_ancestor.md]: There is a typo, I mean O(n) - O(1) LCA...

File: src/tree/lowest_common_ancestor.md (hunk 6 above)
Comment (harry900831) [reply src/tree/lowest_common_ancestor.md]: So why 0.5 log N not K = logN?

File: src/tree/lowest_common_ancestor.md [hunk 14]
```diff
@@ -0,0 +1,928 @@
...
+>
+> - \\( N \leq 2 \times 10^5 \\)
+>
+> - \\( N-1 \leq M \leq 2 \times 10^5 \\)
+
+<details><summary> Solution </summary>
+
+根據題意，我們可以分成兩個 Case 來討論。對於 Case 1，如果給定的邊已經位於該圖的最小生成樹上，那麼我們就直接輸出最小生成樹的權重。而對於 Case 2，如果給定的邊不位於最小生成樹上，那麼我們就需要先將給定的邊加入到最小生成樹中，假設給定邊的兩端點為點 \\( u \\) 和點 \\( v \\)。加入的同時，給定邊會和點 \\( u \\) 到 點 \\( v \\) 的最短路徑組合起來形成環。而為了要取 minimal 且維持樹的結構，我們需要移除環上權重最大的邊，這條邊不會是給定的邊，否則給定的邊會符合 Case 1。因此針對 Case 2，我們需要找到點 \\( u \\) 到 點 \\( v \\) 最短路徑上權重最大的邊。
+
+關於為什麼這樣做會給我們 minimal 的結果，有興趣的讀者可以參考本文附上的參考資料。
```
Comment (harry900831) [inline src/tree/lowest_common_ancestor.md:768]: which one

**Review by harry900831 — CHANGES_REQUESTED** [review]
Please resolve the comment by yourself if you fix it. If you want to discuss about it, just directly reply to it.

**Comment by harry900831** [comment]
I've uploaded the score in advance. I'll merge it after it's finalized.

**Comment by didwhddks** [comment]
@harry900831 助教你好，因為我下個星期都會在國外，所以在 23 號前可能沒辦法 finish 所有的 requested changes。關於 grading 的部分，助教可以先進行，我回國之後會繼續完成剩餘的 requested changes。

**Comment by harry900831** [comment]
Got it, I've uploaded the score. Please check it on eeclass.

**Pull request closed without merging**